Kubeadm/dual stack support in 1.21

pacoxu opened the pull request:
part of https://github.com/kubernetes/kubeadm/issues/1612

- [x] prepare: sudo sysctl -w net.ipv6.conf.all.forwarding=1
- [x] init args for dual stack
- [x] how to use single stack
- [x] --apiserver-advertise-address doesn't support dual-stack yet.
- [x] kubelet --node-ip supports multi IP now. (how kubeadm use that?) I need test on it.
- [x] join a dual stack node 
- [x] upgrade behavior changes: ?
- [x] HA configurations （not in this pr）

https://kubernetes.io/docs/concepts/services-networking/dual-stack/

kubeadm dual-stack install log: https://gist.github.com/pacoxu/3fda7ccbccca82a0cd791cad454ae69f

pacoxu commented:
```
apiVersion: kubeadm.k8s.io/v1beta2
featureGates:
  IPv6DualStack: true
kind: ClusterConfiguration
networking:
      podSubnet: 172.30.0.0/16,fefe:ffff:0::/48
      serviceSubnet: 172.31.0.0/16,fefe:ffff:1::/108
```

Need I add a sample configuration file like?

neolit123 commented:
/cc
@kubernetes/sig-cluster-lifecycle 

i think including an example config is fine, v1beta2 is not deprecated and will be around for at least one more year (if not more).

neolit123 commented on content/en/docs/concepts/services-networking/dual-stack.md:
```diff
@@ -70,6 +70,39 @@ An example of an IPv6 CIDR: `fdXY:IJKL:MNOP:15::/64` (this shows the format but
+
+Install kubeadm following the steps from the [Installing Kubeadm](/docs/setup/production-environment/tools/kubeadm/install-kubeadm/) documentation.
+
+Make sure that nodes allow ipv6 forwarding, if not, run `sudo sysctl -w net.ipv6.conf.all.forwarding=1` on every node in the cluster.
```
`IPv6`
please change this in all places in the new diff.

neolit123 commented on content/en/docs/concepts/services-networking/dual-stack.md:
```diff
@@ -70,6 +70,39 @@ An example of an IPv6 CIDR: `fdXY:IJKL:MNOP:15::/64` (this shows the format but
+
+Make sure that nodes allow IPv6 forwarding, if not, run `sudo sysctl -w net.ipv6.conf.all.forwarding=1` on every node in the cluster.
+
+* Notice that `kubeadm upgrade` will change IPv6DualStack to true by default if feature gate is not set in old cluster. However, cluster-cidr and service-cidr modification are not supported.
```
```suggestion
{{< note >}}
`kubeadm upgrade` will change `IPv6DualStack` to true by default if feature gate is not set in old cluster. However, cluster-cidr and service-cidr modification are not supported.
{{< /note >}}
```

neolit123 commented on content/en/docs/concepts/services-networking/dual-stack.md:
````diff
@@ -70,6 +70,39 @@ An example of an IPv6 CIDR: `fdXY:IJKL:MNOP:15::/64` (this shows the format but
+You must pass your `podCidr` and `serviceCidr` value to `kubeadm init`, like so:
+
+```
+kubeadm init --feature-gates IPv6DualStack=true --pod-network-cidr=172.30.0.0/16,fefe:ffff:0::/48 --service-cidr=172.31.0.0/16,fefe:ffff:1::/108
````
i think we should just pass `--config` with the example config we discussed as the primary example target in this guide.
these kubeadm flags are GA / stable, but we recommend config to users since it's versioned.
we can still include examples with the flags as the second option if you'd like.

neolit123 commented on content/en/docs/concepts/services-networking/dual-stack.md:
```diff
@@ -70,6 +70,39 @@ An example of an IPv6 CIDR: `fdXY:IJKL:MNOP:15::/64` (this shows the format but
+
+Currently, `kube-controller-manager` flags `--node-cidr-mask-size-ipv4|--node-cidr-mask-size-ipv6` are setting with the default value. See [enable IPv4/IPv6 dual stack](/docs/concepts/services-networking/dual-stack#enable-ipv4ipv6-dual-stack).
+
+There is a limitation here, `--apiserver-advertise-address` flag doesn't support dual-stack.
```
we should first mention the config fields:
InitConfiguration -> localAPIEndpoint
and 
JoinConfiguration -> controlPlane -> localAPIEndpoint

https://pkg.go.dev/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta2

we can also mention the flag as they are GA, it's just that we recommend the config to users since it's versioned.

pacoxu replied:
Done

neolit123 commented on content/en/docs/concepts/services-networking/dual-stack.md:
```diff
@@ -70,6 +70,39 @@ An example of an IPv6 CIDR: `fdXY:IJKL:MNOP:15::/64` (this shows the format but
+
+Before joining a node, make sure that the node has IPv6 routable network interface and allows IPv6 forwarding.
+
+
```
```suggestion
```

neolit123 commented on content/en/docs/concepts/services-networking/dual-stack.md:
```diff
@@ -70,6 +70,39 @@ An example of an IPv6 CIDR: `fdXY:IJKL:MNOP:15::/64` (this shows the format but
+
+#### Create a single-stack cluster
+
+As IPv6DualStack feature gate is beta and default to true after 1.21, dual stack will be enabled by default without `--feature-gates` flag.
```
```suggestion
In 1.21 the `IPv6DualStack` feature is Beta and the feature gate is defaulted to `true`. To disable the feature you must configure the feature gate to `false`. Note that once the feature is GA, the feature gate will be removed.
```

pacoxu commented:
- [x]  ipv6 -> IPv6
- [x]  add an example config yaml

neolit123 commented on content/en/docs/concepts/services-networking/dual-stack.md:
````diff
@@ -70,6 +70,39 @@ An example of an IPv6 CIDR: `fdXY:IJKL:MNOP:15::/64` (this shows the format but
+As IPv6DualStack feature gate is beta and default to true after 1.21, dual stack will be enabled by default without `--feature-gates` flag.
+
+```
+kubeadm init --feature-gates IPv6DualStack=false
````
we can have the config snippet again here, but the flag version only is fine too.

pacoxu replied:
Done

neolit123 commented on content/en/docs/concepts/services-networking/dual-stack.md:
```diff
@@ -70,6 +70,39 @@ An example of an IPv6 CIDR: `fdXY:IJKL:MNOP:15::/64` (this shows the format but
 
 {{< /note >}}
 
+### Enable dual-stack via kubeadm
```
i'm not sure if this is the best place for this, actually.

maybe this page can just cross link to a new section under the https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/

sig-docs will likely provide feedback on this.

neolit123 replied:
my vote here is to move this as a new page under:
https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/
and cross reference to this dual-stack.md page.

but we should wait for sig-docs feedback.

pacoxu replied:
> wait for sig-docs feedback

OK

sftim replied:
For SIG Docs usually try _not_ to put task descriptions in the concepts section. We'd prefer to either create a new task page or add a section to an existing task page.

Sometimes https://kubernetes.io/docs/setup/ is a better home; that's usually for documents that only apply to brand new clusters.

neolit123 commented on content/en/docs/concepts/services-networking/dual-stack.md:
```diff
@@ -70,6 +70,57 @@ An example of an IPv6 CIDR: `fdXY:IJKL:MNOP:15::/64` (this shows the format but
+featureGates:
+  IPv6DualStack: true
+networking:
+      podSubnet: 172.30.0.0/16,fefe:ffff:0::/48
+      serviceSubnet: 172.31.0.0/16,fefe:ffff:1::/108
```
```suggestion
  podSubnet: 172.30.0.0/16,fefe:ffff:0::/48
  serviceSubnet: 172.31.0.0/16,fefe:ffff:1::/108
```

neolit123 commented on content/en/docs/concepts/services-networking/dual-stack.md:
````diff
@@ -70,6 +70,57 @@ An example of an IPv6 CIDR: `fdXY:IJKL:MNOP:15::/64` (this shows the format but
+kubeadm init --feature-gates IPv6DualStack=true --pod-network-cidr=172.30.0.0/16,fefe:ffff:0::/48 --service-cidr=172.31.0.0/16,fefe:ffff:1::/108
+```
+
+To make things more clear, here is an example config `kubeadm-config.yml` for the dual-stack control-plane node.
````
```suggestion
To make things more clear, here is an example kubeadm [configuration file](https://pkg.go.dev/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta2) `kubeadm-config.yml` for the dual-stack control plane node.
```

someone told us to not have the dash for `control plane` in docs.
[1]

neolit123 commented on content/en/docs/concepts/services-networking/dual-stack.md:
```diff
@@ -70,6 +70,57 @@ An example of an IPv6 CIDR: `fdXY:IJKL:MNOP:15::/64` (this shows the format but
+      serviceSubnet: 172.31.0.0/16,fefe:ffff:1::/108
+``
+
+Run kubeadm to initiate the dual-stack control-plane node.
```
[1]

neolit123 reviewed (commented):
seems fine to me, not sure if this should be the page where we add it.

most importantly the PR should be against k/website's dev-1.21 branch.
the "change branch" option of github usually fails for k/website.

you can try:
- change the PR branch here
- do `git diff ... > patch.diff` locally to store the diff
- delete local branch
- create a branch with same name based on `dev-1.21` locally
- apply the patch
- push to the remote

neolit123 commented:
@ sig-docs / @ sig -network do you think that we should have this guide in the kubeadm docs and just cross link to it from the sig network dual-stack page?

pacoxu commented:
Thanks @neolit123. It works.
I changed the codebase to 1.21 and update the PR target branch.

pacoxu commented:
/sig network cluster-lifecycle
/kind feature 
/area kubeadm
/milestone v1.21

neolit123 commented on content/en/docs/concepts/services-networking/dual-stack.md:
````diff
@@ -70,6 +70,112 @@ An example of an IPv6 CIDR: `fdXY:IJKL:MNOP:15::/64` (this shows the format but
+```
+
+Besides, here is an example kubeadm [configuration file](https://pkg.go.dev/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta2) `kubeadm-config.yml` for joining another control plane node to the cluster.
+```yaml
````
```suggestion

```yaml
```

neolit123 commented on content/en/docs/concepts/services-networking/dual-stack.md:
```diff
@@ -70,6 +70,112 @@ An example of an IPv6 CIDR: `fdXY:IJKL:MNOP:15::/64` (this shows the format but
+
+#### Create a dual-stack cluster
+
+A simple command with `podCidr` and `serviceCidr` flags to create a dual-stack cluster via `kubeadm init`, it like below:
```
```suggestion
A simple command with `--pod-network-cidr` and `--service-cidr` flags to create a dual-stack cluster via `kubeadm init`, it like below:
```

neolit123 commented:
this looks mostly fine to me, thanks.
let's wait on SIG Docs / Network for feedback on:
https://github.com/kubernetes/website/pull/26675#issuecomment-785532901
https://github.com/kubernetes/website/pull/26675#discussion_r582445493

reylejano commented:
/milestone 1.21
/assign @mvortizr 
/hold until [k/k PR 99294](https://github.com/kubernetes/kubernetes/pull/99294) merges

sftim commented on content/en/docs/concepts/services-networking/dual-stack.md:
```diff
@@ -70,6 +70,113 @@ An example of an IPv6 CIDR: `fdXY:IJKL:MNOP:15::/64` (this shows the format but
+`kubeadm upgrade` will change `IPv6DualStack` to true by default if feature gate is not set in old cluster. However, cluster-cidr and service-cidr modification are not supported.
+{{< /note >}}
+
+#### Create a dual-stack cluster
```
Perhaps this could live inside https://kubernetes.io/docs/setup/

pacoxu replied:
Moved

neolit123 commented on content/en/docs/setup/production-environment/tools/kubeadm/enable-disable-dual-stack.md:
```diff
@@ -0,0 +1,130 @@
+- sig-cluster-lifecycle
+- neolit123
+- sftim
+title: Enable Or Disable Dual-Stack via kubeadm
```
ok, give this is a feature gate it means that the feature at some point will be locked to true and cannot be disabled.
with that in mind, IMO we should name the page more broadly:

```suggestion
title: Dual-stack support with kubeadm
```

```
dual-stack-support.md
```

once the feature is GA, we can keep some of the details, but remove the feature gate related toggles.

aojea commented on content/en/docs/setup/production-environment/tools/kubeadm/enable-disable-dual-stack.md:
````diff
@@ -0,0 +1,130 @@
+A simple command with `podCidr` and `serviceCidr` flags to create a dual-stack cluster via `kubeadm init`, it like below:
+
+```shell
+kubeadm init --feature-gates IPv6DualStack=true --pod-network-cidr=172.30.0.0/16,fefe:ffff:0::/48 --service-cidr=172.31.0.0/16,fefe:ffff:1::/108
````
the ipv6 mask /108 for services is not allowed, max 112, also people use to take the ip addresses for the documentation literally so I suggest to use better the common values we have in KIND for dual-stack

```
pod-network-cidr=10.244.0.0/16,fd00:10:244::/56
service-cidr=10.96.0.0/16,fd00:10:96::/112
```

aojea commented on content/en/docs/setup/production-environment/tools/kubeadm/enable-disable-dual-stack.md:
```diff
@@ -0,0 +1,130 @@
+featureGates:
+  IPv6DualStack: true
+networking:
+  podSubnet: 172.30.0.0/16,fefe:ffff:0::/48
+  serviceSubnet: 172.31.0.0/16,fefe:ffff:1::/108
```
ditto https://github.com/kubernetes/website/pull/26675/files#r584168282

pacoxu replied:
Done

aojea commented on content/en/docs/setup/production-environment/tools/kubeadm/enable-disable-dual-stack.md:
````diff
@@ -0,0 +1,130 @@
+localAPIEndpoint:
+  advertiseAddress: "10.100.0.1"
+  bindPort: 6443
+```
````
kubelet need to be configured with the nodeIPs of each IP family

```
nodeRegistration:
  kubeletExtraArgs:
    fail-swap-on: "false"
    node-ip: 10.100.0.2,fd00:1:2:3::2
 ```

aojea commented on content/en/docs/setup/production-environment/tools/kubeadm/enable-disable-dual-stack.md:
````diff
@@ -0,0 +1,130 @@
+kubeadm join --config=kubeadm-config.yml ...
+```
+
+### Create a single-stack cluster
````
Enabling the dual-stack feature doesn't mean that you need to use dual-stack addresses, i.e., you can have a single-stack cluster with the feature flag enabled ... I think current validation logic allows that but I need to double check
@neolit123 ^

neolit123 replied:
at k/k master we merged the feature gate as beta (enabled=true) and our e2e tests are passing where only IPv4 addresses are used. https://k8s-testgrid.appspot.com/sig-cluster-lifecycle-kubeadm#kubeadm-kinder-latest

pacoxu replied:
Add a note

neolit123 commented on content/en/docs/setup/production-environment/tools/kubeadm/enable-disable-dual-stack.md:
````diff
@@ -0,0 +1,130 @@
+kubeadm init --feature-gates IPv6DualStack=true --pod-network-cidr=172.30.0.0/16,fefe:ffff:0::/48 --service-cidr=172.31.0.0/16,fefe:ffff:1::/108
+```
+
+To make things more clear, here is an example kubeadm [configuration file](https://pkg.go.dev/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta2) `kubeadm-config.yml` for the dual-stack control plane node.
````
```suggestion
To make things more clear, here is an example kubeadm [configuration file](https://pkg.go.dev/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta2) `kubeadm-config.yaml` for the dual-stack control plane node.
```

we are merging a pr for k/k/cmd/kubeadm to use YAML everywhere.

neolit123 replied:
please apply this to the rest of the example filenames in the doc here.

pacoxu replied:
Done

neolit123 commented on content/en/docs/setup/production-environment/tools/kubeadm/enable-disable-dual-stack.md:
````diff
@@ -0,0 +1,130 @@
+Run kubeadm to initiate the dual-stack control plane node.
+
+```shell
+kubeadm init --config=kubeadm-config.yml
````
```suggestion
kubeadm init --config=kubeadm-config.yaml
```

pacoxu replied:
Done

aojea commented on content/en/docs/setup/production-environment/tools/kubeadm/enable-disable-dual-stack.md:
````diff
@@ -0,0 +1,138 @@
+
+```yaml
+apiVersion: kubeadm.k8s.io/v1beta2
+kind: JoinConfiguration
````
you need to set the kubelet node-ip dual-stack addresses here too IIRC

pacoxu replied:
Updated.
Missing it here.

neolit123 commented on content/en/docs/setup/production-environment/tools/kubeadm/enable-disable-dual-stack.md:
```diff
@@ -0,0 +1,147 @@
+- sftim
+title: Enable Or Disable Dual-Stack via kubeadm
+feature:
+  title: Enable Or Disable Dual-Stack via kubeadm
```
as mentioned here we should make the title more generic:
> Dual-stack support with kubeadm

https://github.com/kubernetes/website/pull/26675/files#r583667883

and file rename `dual-stack-support.md`

neolit123 commented on content/en/docs/setup/production-environment/tools/kubeadm/enable-disable-dual-stack.md:
```diff
@@ -0,0 +1,147 @@
+
+## {{% heading "whatsnext" %}}
+
+
```
```suggestion
```

neolit123 commented on content/en/docs/setup/production-environment/tools/kubeadm/enable-disable-dual-stack.md:
```diff
@@ -0,0 +1,147 @@
+---
+reviewers:
+- sig-cluster-lifecycle
+- neolit123
+- sftim
```
```suggestion
```

we should remove this.

neolit123 commented on content/en/docs/setup/production-environment/tools/kubeadm/enable-disable-dual-stack.md:
```diff
@@ -0,0 +1,147 @@
+Make sure that nodes allow IPv6 forwarding, if not, run `sudo sysctl -w net.ipv6.conf.all.forwarding=1` on every node in the cluster.
+
+{{< note >}}
+`kubeadm upgrade` will change `IPv6DualStack` to true by default if feature gate is not set in old cluster. However, cluster-cidr and service-cidr modification are not supported.
```
```suggestion
`kubeadm upgrade` will change `IPv6DualStack` to true by default if the feature gate is not set in an old cluster. However, cluster-cidr and service-cidr modification are not supported.
```

neolit123 commented on content/en/docs/setup/production-environment/tools/kubeadm/enable-disable-dual-stack.md:
```diff
@@ -0,0 +1,147 @@
+
+### Create a dual-stack cluster
+
+A simple command with `podCidr` and `serviceCidr` flags to create a dual-stack cluster via `kubeadm init`, it like below:
```
```suggestion
To create a dual-stack cluster with `kubeadm init` you can pass the following flags:
```

neolit123 commented on content/en/docs/setup/production-environment/tools/kubeadm/enable-disable-dual-stack.md:
````diff
@@ -0,0 +1,147 @@
+A simple command with `podCidr` and `serviceCidr` flags to create a dual-stack cluster via `kubeadm init`, it like below:
+
+```shell
+kubeadm init --feature-gates IPv6DualStack=true --pod-network-cidr=10.244.0.0/16,fd00:10:244::/56 --service-cidr=10.96.0.0/16,fd00:10:96::/112
````
```suggestion
kubeadm init --pod-network-cidr=10.244.0.0/16,fd00:10:244::/56 --service-cidr=10.96.0.0/16,fd00:10:96::/112
```

since the FG becomes beta in 1.21 we don't need to pass `--feature-gates IPv6DualStack=true`

neolit123 commented on content/en/docs/setup/production-environment/tools/kubeadm/enable-disable-dual-stack.md:
````diff
@@ -0,0 +1,147 @@
+kubeadm init --feature-gates IPv6DualStack=true --pod-network-cidr=10.244.0.0/16,fd00:10:244::/56 --service-cidr=10.96.0.0/16,fd00:10:96::/112
+```
+
+To make things more clear, here is an example kubeadm [configuration file](https://pkg.go.dev/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta2) `kubeadm-config.yaml` for the dual-stack control plane node.
````
```suggestion
To make things clearer, here is an example kubeadm [configuration file](https://pkg.go.dev/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta2) `kubeadm-config.yaml` for the primary dual-stack control plane node.
```

neolit123 commented on content/en/docs/setup/production-environment/tools/kubeadm/enable-disable-dual-stack.md:
````diff
@@ -0,0 +1,147 @@
+    node-ip: 10.100.0.2,fd00:1:2:3::2
+```
+
+`advertiseAddress` in InitConfiguration specifies the IP address the API Server will advertise it's listening on. It equals to `--apiserver-advertise-address` flag of `kubeadm init`.
````
```suggestion
`advertiseAddress` in InitConfiguration specifies the IP address the API Server will advertise it's listening on. It equals to the `--apiserver-advertise-address` flag of `kubeadm init`.
```

neolit123 commented on content/en/docs/setup/production-environment/tools/kubeadm/enable-disable-dual-stack.md:
````diff
@@ -0,0 +1,147 @@
+kubeadm init --config=kubeadm-config.yaml
+```
+
+Currently, `kube-controller-manager` flags `--node-cidr-mask-size-ipv4|--node-cidr-mask-size-ipv6` are setting with the default value. See [enable IPv4/IPv6 dual stack](/docs/concepts/services-networking/dual-stack#enable-ipv4ipv6-dual-stack).
````
```suggestion
Currently, the kube-controller-manager flags `--node-cidr-mask-size-ipv4|--node-cidr-mask-size-ipv6` are being left with default values. See [enable IPv4/IPv6 dual stack](/docs/concepts/services-networking/dual-stack#enable-ipv4ipv6-dual-stack).
```

^ not sure if my suggestion is correct in terms of context.

neolit123 commented on content/en/docs/setup/production-environment/tools/kubeadm/enable-disable-dual-stack.md:
```diff
@@ -0,0 +1,147 @@
+
+Currently, `kube-controller-manager` flags `--node-cidr-mask-size-ipv4|--node-cidr-mask-size-ipv6` are setting with the default value. See [enable IPv4/IPv6 dual stack](/docs/concepts/services-networking/dual-stack#enable-ipv4ipv6-dual-stack).
+
+There is a limitation here, `--apiserver-advertise-address` flag doesn't support dual-stack.
```
```suggestion
{{< note >}}
Currently the `--apiserver-advertise-address` flag does not support dual-stack.
{{< /note >}}
```

why does `apiserver-advertise-address` does not support dual stack?
what else can we add here as context in terms of the core components?

aojea replied:
this has a lot of implications and is not a clear win to dual-stack it, neither it has a good use case backing it up

neolit123 commented on content/en/docs/setup/production-environment/tools/kubeadm/enable-disable-dual-stack.md:
```diff
@@ -0,0 +1,147 @@
+
+Before joining a node, make sure that the node has IPv6 routable network interface and allows IPv6 forwarding.
+
+
```
```suggestion
```

neolit123 commented on content/en/docs/setup/production-environment/tools/kubeadm/enable-disable-dual-stack.md:
````diff
@@ -0,0 +1,147 @@
+    node-ip: 10.100.0.3,fd00:1:2:3::3
+```
+
+Besides, here is an example kubeadm [configuration file](https://pkg.go.dev/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta2) `kubeadm-config.yaml` for joining another control plane node to the cluster.
````
```suggestion
Also, here is an example kubeadm [configuration file](https://pkg.go.dev/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta2) `kubeadm-config.yaml` for joining another control plane node to the cluster.
```

neolit123 commented on content/en/docs/setup/production-environment/tools/kubeadm/enable-disable-dual-stack.md:
````diff
@@ -0,0 +1,147 @@
+
+```
+
+`advertiseAddress` in JoinConfiguration.controlPlane specifies the IP address the API Server will advertise it's listening on. It equals to `--apiserver-advertise-address` flag of `kubeadm join`.
````
```suggestion
`advertiseAddress` in JoinConfiguration.controlPlane specifies the IP address the API Server will advertise it's listening on. It equals to the `--apiserver-advertise-address` flag of `kubeadm join`.
```

pacoxu commented:
Updated. 
Thanks for suggestions.

neolit123 commented:
/lgtm
/hold cancel

LGTM, for the tech review. thanks
sig-docs might request more changes to the language / wording.

aojea commented:
LGTM

pacoxu commented:
/assign @jimangel

pacoxu commented:
/cc @sftim 
for approval

sftim commented on content/en/docs/setup/production-environment/tools/kubeadm/dual-stack-support.md:
```diff
@@ -0,0 +1,144 @@
+---
+title: Dual-stack support with kubeadm
```
The improvement is dual stack support, but once that lands it's just “dual stack” (think about the page about RBAC; once it was new, but the page about it doesn't talk about RBAC support, just about RBAC).

How about:
```suggestion
title: Enable dual-stack networking using kubeadm
```
?

pacoxu replied:
@neolit123 Do you agree to change the title?

`Dual-stack support with kubeadm` is simple.
I prefer `Enable dual-stack networking using kubeadm` as Tim says. The page mainly tells about how to enable it.

neolit123 replied:
i don't object.

sftim commented on content/en/docs/setup/production-environment/tools/kubeadm/dual-stack-support.md:
```diff
@@ -0,0 +1,144 @@
+
+{{< feature-state for_k8s_version="v1.21" state="beta" >}}
+
+IPv4/IPv6 dual-stack enables the allocation of both IPv4 and IPv6 addresses to {{< glossary_tooltip text="Pods" term_id="pod" >}} and {{< glossary_tooltip text="Services" term_id="service" >}}.
```
How about
```suggestion
Your Kubernetes cluster can run in [dual-stack](/docs/concepts/services-networking/dual-stack/) networking mode, which means that cluster networking lets you use either address family. In a dual-stack cluster, the control plane can assign both an IPv4 address and an IPv6 address to a single {{< glossary_tooltip text="Pod" term_id="pod" >}} or a {{< glossary_tooltip text="Service" term_id="service" >}}.
```

sftim commented on content/en/docs/setup/production-environment/tools/kubeadm/dual-stack-support.md:
```diff
@@ -0,0 +1,144 @@
+---
+title: Dual-stack support with kubeadm
+feature:
+  title: Dual-stack support with kubeadm
```
Let's hold off on advertising that as a feature until it's GA (or at least until we can guarantee in which release it will graduate to stable).

sftim commented on content/en/docs/setup/production-environment/tools/kubeadm/dual-stack-support.md:
```diff
@@ -0,0 +1,144 @@
+
+<!-- body -->
+
+## Enable dual-stack via kubeadm
+
+Install kubeadm following the steps from the [Installing Kubeadm](/docs/setup/production-environment/tools/kubeadm/install-kubeadm/) documentation.
+
+Make sure that nodes allow IPv6 forwarding, if not, run `sudo sysctl -w net.ipv6.conf.all.forwarding=1` on every node in the cluster.
```
I'd change both the formatting and the advice:
```suggestion
## {{% heading "prerequisites" %}}

You need to have installed the {{< glossary_tooltip text="kubeadm" term_id="kubeadm" >}} tool, following the steps from [Installing kubeadm](/docs/setup/production-environment/tools/kubeadm/install-kubeadm/).

For each server that you want to use as a {{< glossary_tooltip text="node" term_id="node" >}}, make sure it allows IPv6 forwarding. On Linux, you can set this by running run `sysctl -w net.ipv6.conf.all.forwarding=1` as the root user on each server.

You need to have an IPv4 and and IPv6 address range to use. Cluster operators typically
use private address ranges for IPv4. For IPv6, a cluster operator typically chooses a global
unicast address block from within `2000::/3`, using a range that is assigned to the operator.  
You don't have to route the cluster's IP address ranges to the public internet.

The size of the IP address allocations should be suitable for the number of Pods and
Services that you are planning to run.
```

Notice that I'm suggesting using global unicast IPv6 ranges. That matches what I see as common practice
for IPv6 adoption generally.

sftim commented on content/en/docs/setup/production-environment/tools/kubeadm/dual-stack-support.md:
```diff
@@ -0,0 +1,144 @@
+Make sure that nodes allow IPv6 forwarding, if not, run `sudo sysctl -w net.ipv6.conf.all.forwarding=1` on every node in the cluster.
+
+{{< note >}}
+`kubeadm upgrade` will change `IPv6DualStack` to true by default if the feature gate is not set in an old cluster. However, cluster CIDR and service CIDR modifications are not supported.
```
I suggest:
```suggestion
If you are upgrading an existing cluster then, by default, the `kubeadm upgrade` command
changes the [feature gate](/docs/reference/command-line-tools-reference/feature-gates/)
`IPv6DualStack` to `true` if that is not already enabled.  
However, `kubeadm` does not support making modifications to the pod IP address range
(“cluster CIDR”) nor to the cluster's Service address range (“Service CIDR”).
```

sftim commented on content/en/docs/setup/production-environment/tools/kubeadm/dual-stack-support.md:
````diff
@@ -0,0 +1,144 @@
+To create a dual-stack cluster with `kubeadm init` you can pass the following flags:
+
+```shell
+kubeadm init --pod-network-cidr=10.244.0.0/16,fd00:10:244::/56 --service-cidr=10.96.0.0/16,fd00:10:96::/112
````
I'd use a documentation address range for IPv6. The range `fd00::/8` is deprecated.

```suggestion
# These address ranges are examples
kubeadm init --pod-network-cidr=10.244.0.0/16,2001:db8:42:0::/56 --service-cidr=10.96.0.0/16,2001:db8:42:1::/112
```

sftim commented on content/en/docs/setup/production-environment/tools/kubeadm/dual-stack-support.md:
````diff
@@ -0,0 +1,144 @@
+To make things clearer, here is an example kubeadm [configuration file](https://pkg.go.dev/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta2) `kubeadm-config.yaml` for the primary dual-stack control plane node.
+
+```yaml
+apiVersion: kubeadm.k8s.io/v1beta2
````
As there are multiple documents in this YAML stream, I'd write:
```suggestion
---
apiVersion: kubeadm.k8s.io/v1beta2
```

sftim commented on content/en/docs/setup/production-environment/tools/kubeadm/dual-stack-support.md:
```diff
@@ -0,0 +1,144 @@
+
+`advertiseAddress` in InitConfiguration specifies the IP address the API Server will advertise it's listening on. It equals to the `--apiserver-advertise-address` flag of `kubeadm init`.
+
+Run kubeadm to initiate the dual-stack control plane node.
```
```suggestion
Run kubeadm to initiate the dual-stack control plane node:
```

sftim commented on content/en/docs/setup/production-environment/tools/kubeadm/dual-stack-support.md:
```diff
@@ -0,0 +1,144 @@
+Currently, the kube-controller-manager flags `--node-cidr-mask-size-ipv4|--node-cidr-mask-size-ipv6` are being left with default values. See [enable IPv4/IPv6 dual stack](/docs/concepts/services-networking/dual-stack#enable-ipv4ipv6-dual-stack).
+
+{{< note >}}
+Currently the `--apiserver-advertise-address` flag does not support dual-stack.
```
We usually don't write “currently” etc unless revising the behavior is a firm commitment.
```suggestion
The `--apiserver-advertise-address` flag does not support dual-stack.
```

sftim commented on content/en/docs/setup/production-environment/tools/kubeadm/dual-stack-support.md:
````diff
@@ -0,0 +1,144 @@
+
+```
+
+`advertiseAddress` in JoinConfiguration.controlPlane specifies the IP address the API Server will advertise it's listening on. It equals to the `--apiserver-advertise-address` flag of `kubeadm join`.
````
nit:
```suggestion
`advertiseAddress` in JoinConfiguration.controlPlane specifies the IP address that the API Server will advertise it is listening on. The value of `advertiseAddress` equals the `--apiserver-advertise-address` flag of `kubeadm join`.
```

sftim commented on content/en/docs/setup/production-environment/tools/kubeadm/dual-stack-support.md:
```diff
@@ -0,0 +1,144 @@
+### Create a single-stack cluster
+
+{{< note >}}
+Enabling the dual-stack feature doesn't mean that you need to use dual-stack addresses, i.e., you can have a single-stack cluster with the feature flag enabled.
```
How about:
```suggestion
Enabling the dual-stack feature doesn't mean that you need to use dual-stack addressing.  
You can deploy a single-stack cluster that has the dual-stack networking feature enabled.
```

sftim commented on content/en/docs/setup/production-environment/tools/kubeadm/dual-stack-support.md:
```diff
@@ -0,0 +1,144 @@
+## {{% heading "whatsnext" %}}
+
+* [Validate IPv4/IPv6 dual-stack](/docs/tasks/network/validate-dual-stack) networking
+* [Dual-stack enabled](/docs/concepts/services-networking/dual-stack/) cluster
```
```suggestion
* Read about [Dual-stack](/docs/concepts/services-networking/dual-stack/) cluster networking
```

sftim commented on content/en/docs/concepts/services-networking/dual-stack.md:
```diff
@@ -242,3 +242,4 @@ Ensure your {{< glossary_tooltip text="CNI" term_id="cni" >}} provider supports
 
 
 * [Validate IPv4/IPv6 dual-stack](/docs/tasks/network/validate-dual-stack) networking
+* [Dual-stack support with kubeadm](/docs/setup/production-environment/tools/kubeadm/dual-stack-support/)
```
```suggestion
* [Enable dual-stack networking using kubeadm
](/docs/setup/production-environment/tools/kubeadm/dual-stack-support/)
```
(if making the suggested change to the title of that page)

However, this is just an example. Really I'd prefer to see “support” removed from the filename and generated URL.

sftim commented on content/en/docs/setup/production-environment/tools/kubeadm/dual-stack-support.md:
```diff
@@ -0,0 +1,144 @@
+  title: Dual-stack support with kubeadm
+content_type: task
+weight: 110
+---
```
I'd also add:
```suggestion
min-kubernetes-server-version: 1.21
---
```

sftim commented on content/en/docs/setup/production-environment/tools/kubeadm/dual-stack-support.md:
````diff
@@ -0,0 +1,144 @@
+    node-ip: 10.100.0.2,fd00:1:2:3::2
+```
+
+`advertiseAddress` in InitConfiguration specifies the IP address the API Server will advertise it's listening on. It equals to the `--apiserver-advertise-address` flag of `kubeadm init`.
````
```suggestion
`advertiseAddress` in InitConfiguration specifies the IP address that the API Server will advertise it is listening on. The value of `advertiseAddress` equals the `--apiserver-advertise-address` flag of `kubeadm init`
```

sftim commented on content/en/docs/setup/production-environment/tools/kubeadm/dual-stack-support.md:
```diff
@@ -0,0 +1,144 @@
+
+### Create a dual-stack cluster
+
+To create a dual-stack cluster with `kubeadm init` you can pass the following flags:
```
I suggest
```suggestion
To create a dual-stack cluster with `kubeadm init` you can pass command line arguments
similar to the following example:
```
```

sftim commented on content/en/docs/setup/production-environment/tools/kubeadm/dual-stack-support.md:
```diff
@@ -0,0 +1,144 @@
+featureGates:
+  IPv6DualStack: true
+networking:
+  podSubnet: 10.244.0.0/16,fd00:10:244::/56
+  serviceSubnet: 10.96.0.0/16,fd00:10:96::/112
```
```suggestion
  podSubnet: 10.244.0.0/16,2001:db8:42:0::/56     # example ranges
  serviceSubnet: 10.96.0.0/16,2001:db8:42:1::/112 # example ranges
```

aojea replied:
hehe, that is my fault, those are the default kind valudes, Tim is always spot on these things, docs recommended ranges are 2001:db8

sftim commented on content/en/docs/setup/production-environment/tools/kubeadm/dual-stack-support.md:
```diff
@@ -0,0 +1,144 @@
+    caCertHashes: ["sha256:fcb3e956a6880c05fc9d09714424b827f57a6fdc8afc44497180905946527adf"]
+nodeRegistration:
+  kubeletExtraArgs:
+    fail-swap-on: "false"
```
Why set `fail-swap-on`?

neolit123 replied:
yep, this can be omitted.
we can also keep the config minimal without non-relevant values like the kubelet extra args and CA-hash / tokens.

pacoxu replied:
removed.

sftim reviewed (commented):
Thanks for proposing this new page.

- I'd prefer not to recommend using `fd00::/8` as this range is already deprecated. Let's use an IPv6 documentation IPv6 range instead. I added an example inline.
- If it's feasible, lets use port 443 as the HTTPS port. This matches typical port assignments on the wider internet.
- I'm not sure why this page recommends setting `fail-swap-on` to `false`.

aojea commented on content/en/docs/concepts/services-networking/dual-stack.md:
```diff
@@ -70,6 +70,113 @@ An example of an IPv6 CIDR: `fdXY:IJKL:MNOP:15::/64` (this shows the format but
+  IPv6DualStack: true
+networking:
+  podSubnet: 172.30.0.0/16,fefe:ffff:0::/48
+  serviceSubnet: 172.31.0.0/16,fefe:ffff:1::/108
```
lowercase is the rfc recommended, and service subnet max size is 112 IIRC

pacoxu replied:
podSubnet: 10.244.0.0/16,2001:db8:42:0::/56
  serviceSubnet: 10.96.0.0/16,2001:db8:42:1::/112

Updated

neolit123 commented on content/en/docs/concepts/services-networking/dual-stack.md:
```diff
@@ -70,6 +70,113 @@ An example of an IPv6 CIDR: `fdXY:IJKL:MNOP:15::/64` (this shows the format but
+kind: InitConfiguration
+localAPIEndpoint:
+  advertiseAddress: "10.100.0.1"
+  bindPort: 6443
```
i don't object to this, but for kubeadm users are more used to 6443, because that is the default.

pacoxu replied:
So I will keep it as 6443 or remove it ?

neolit123 replied:
my vote is to leave it 6443.

neolit123 commented on content/en/docs/concepts/services-networking/dual-stack.md:
````diff
@@ -70,6 +70,113 @@ An example of an IPv6 CIDR: `fdXY:IJKL:MNOP:15::/64` (this shows the format but
+kubeadm init --feature-gates IPv6DualStack=false
+```
+
+To make things more clear, here is an example kubeadm [configuration file](https://pkg.go.dev/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta2) `kubeadm-config.yml` for the single-stack control plane node.
````
+1, emphasizing that is a good idea.

pacoxu replied:
```
Enabling the dual-stack feature doesn't mean that you need to use dual-stack addressing.  
You can deploy a single-stack cluster that has the dual-stack networking feature enabled.
```

Current description. Need I add more?

neolit123 replied:
yes, seems good.

pacoxu commented:
> Thanks for proposing this new page.
> 
> * I'd prefer not to recommend using `fd00::/8` as this range is already deprecated. Let's use an IPv6 documentation IPv6 range instead. I added an example inline.
> * If it's feasible, lets use port 443 as the HTTPS port. This matches typical port assignments on the wider internet.
> * I'm not sure why this page recommends setting `fail-swap-on` to `false`.

@sftim  Thanks for your reveiw.
Updated and rebased, except the HTTPS port as Lubomir commented that it is a default value from kubeadm.

neolit123 commented:
/lgtm

reylejano commented:
Hi @pacoxu , this PR has a tech lgtm and updates addressing comments from a sig docs reviewer. 
@kubernetes/sig-docs-pr-reviews Can you review the updates and give an lgtm?

Also wanted to share upcoming doc related dates for the 1.21 release:
- <strike> Docs Ready for Review deadline is March 24</strike> done
- Docs Ready to Merge deadline is March 31

reylejano commented:
@pacoxu Can you squash your commits

Thank you

pacoxu commented:
@reylejano updated.

tengqm commented:
/approve
